lib/nostrum/voice/crypto.ex

```diff
@@ -0,0 +1,223 @@
+
+    {unpadded_nonce, nonce} = lite_nonce(voice)
+
+    [header, Salsa.encrypt(data, key, nonce), unpadded_nonce]
+  end
+
+  def encrypt_xsalsa20_poly1305_lite_rtpsize(voice, data),
+    do: encrypt_xsalsa20_poly1305_lite(voice, data)
+
+  def encrypt_xchacha20_poly1305(voice, data),
+    do: encrypt_aead_xchacha20_poly1305_rtpsize(voice, data)
+
+  def encrypt_aead_xchacha20_poly1305_rtpsize(%VoiceState{secret_key: key} = voice, data) do
+    header = Audio.rtp_header(voice)
+
+    {unpadded_nonce, nonce} = lite_nonce(voice)
+
+    [header, Chacha.encrypt(data, key, nonce, _aad = header), unpadded_nonce]
+  end
+
+  def encrypt_aead_aes256_gcm(voice, data), do: encrypt_aes256_gcm(voice, data)
+  def encrypt_aead_aes256_gcm_rtpsize(voice, data), do: encrypt_aes256_gcm(voice, data)
+
+  def encrypt_aes256_gcm(%VoiceState{secret_key: key} = voice, data) do
+    header = Audio.rtp_header(voice)
+
+    {unpadded_nonce, nonce} = lite_nonce(voice, 12)
+
+    [header, Aes.encrypt(data, key, nonce, _aad = header), unpadded_nonce]
+  end
+
+  def decrypt_xsalsa20_poly1305(key, <<header::bytes-size(12), cipher_text::binary>>) do
+    nonce = header <> <<0::unit(8)-size(12)>>
+
+    Salsa.decrypt(cipher_text, key, nonce)
+  end
+
+  def decrypt_xsalsa20_poly1305_lite(key, data) do
+    {_header, cipher_text, _tag = <<>>, nonce} = decode_packet(data, 4, 24, 0)
+
+    Salsa.decrypt(cipher_text, key, nonce)
+  end
+
+  def decrypt_xsalsa20_poly1305_suffix(key, data) do
+    {_header, cipher_text, _tag = <<>>, nonce} = decode_packet(data, 24, 24, 0)
+
+    Salsa.decrypt(cipher_text, key, nonce)
+  end
+
+  def decrypt_xsalsa20_poly1305_lite_rtpsize(key, data) do
+    {_header, cipher_text, _tag, nonce, ext_len} = decode_packet_rtpsize(data, 24, 0)
+
+    <<_exts::unit(32)-size(ext_len), opus::binary>> = Salsa.decrypt(cipher_text, key, nonce)
+
+    opus
+  end
+
+  def decrypt_xchacha20_poly1305(key, data),
+    do: decrypt_aead_xchacha20_poly1305_rtpsize(key, data)
+
+  def decrypt_aead_xchacha20_poly1305_rtpsize(key, data) do
+    {header, cipher_text, tag, nonce, ext_len} = decode_packet_rtpsize(data, 24)
+
+    <<_exts::unit(32)-size(ext_len), opus::binary>> =
+      Chacha.decrypt(cipher_text, key, nonce, _aad = header, tag)
+
+    opus
+  end
+
+  def decrypt_aes256_gcm(key, data), do: decrypt_aead_aes256_gcm_rtpsize(key, data)
+
+  def decrypt_aead_aes256_gcm_rtpsize(key, data) do
+    {header, cipher_text, tag, nonce, ext_len} = decode_packet_rtpsize(data, 12)
+
+    <<_exts::unit(32)-size(ext_len), opus::binary>> =
+      Aes.decrypt(cipher_text, key, nonce, _aad = header, tag)
+
+    opus
+  end
+
+  def decrypt_aead_aes256_gcm(key, data) do
+    {header, cipher_text, tag, nonce} = decode_packet(data, 4, 12, 16)
+
+    Aes.decrypt(cipher_text, key, nonce, _aad = header, tag)
+  end
+
+  @lite_nonce_length 4
+
+  defp lite_nonce(%VoiceState{rtp_sequence: rtp_sequence}, nonce_length \\ 24) do
+    unpadded_nonce = <<rtp_sequence::32>>
```

Comment:
For 'lite' nonce modes Discord specifies a 32-bit sequential integer, which I am reusing the rtp sequence number for. Discord is doing something similar, thought apparently serializing the sequential in as little endian.

Even though it's 16 bits for the RTP header, the integer we keep in state is just an integer, and the `rtp_header/1` function keeps it bound between 0 and max 16-bit-unsigned while we keep incrementing our internal state to positive infinity.
`<<..., voice.rtp_sequence::16, ...>>`

With respect to that RTP RFC stating the sequence number 'SHOULD be random (unpredictable)' to ward against known-plaintext attacks when encrypted, we start ours at `0`, but since the RTP header is never encrypted it doesn't matter that it's not random. I would be more worried about short packets being easily identified as Opus silence, known plaintext of `<<0xF8, 0xFF, 0xFE>>`

Internally with the initial state of the keystream block for Salsa and ChaCha ciphers, yes the block counter is a part of the input along with the nonce. Interestingly, due to Discord's lite nonce scheme, the last 8 bytes of our 24 byte nonce for XSalsa20 and XChaCha20 that aren't ingested during the subkey generation are ALWAYS all zeros.

Regardless of if the nonces are pseudo-randomly generated, as long as they are never used more than once with the same key, the ciphers are said to be secure. In our use case with 32-bit sequential nonce for `lite` modes and with each voice packet being 20ms in length, we are guaranteed to never reuse a nonce for up to 994 days of continuous speech. Discord will always invalidate a voice session with a 4006 ws error code long before that happens, at which point we would rejoin with a new key.